### ISSUE TITLE
Custom Shaders in r76

### ISSUE THREAD

**Issue by RemusMar** [mrdoob/three.js#8685]
It worked perfect in r71 ... r75:
http://necromanthus.com/Test/html5/Lara_shader.html 
In r76 I don't get any error, but I don't see anything:
http://necromanthus.com/Test/html5/Lara_shader_r76.html 
Bug or ... broken backward compatibility (again) ?

**Comment by mrdoob** [comment]
I'm not able to inspect the examples at the moment, but we don't really guarantee backwards compatibility with custom shaders.

**Comment by RemusMar** [comment]
> we don't really guarantee backwards compatibility with custom shaders. 

Ricardo,
There is always a workaround (for anything).
But if I don't get any error, how do you expect me to debug something like that?

**Comment by mrdoob** [comment]
I'll take a look later tonight.

**Comment by RemusMar** [comment]
The r76 issue is related to Vertex Shader and Skinned Mesh.
		
		#ifdef USE_SKINNING
			mat4 boneMatX = getBoneMatrix( skinIndex.x );
			mat4 boneMatY = getBoneMatrix( skinIndex.y );
			mat4 boneMatZ = getBoneMatrix( skinIndex.z );
			mat4 boneMatW = getBoneMatrix( skinIndex.w );
			mat4 skinMatrix = mat4( 0.0 );
			skinMatrix += skinWeight.x * boneMatX;
			skinMatrix += skinWeight.y * boneMatY;
			skinMatrix += skinWeight.z * boneMatZ;
			skinMatrix += skinWeight.w * boneMatW;
			skinMatrix  = bindMatrixInverse * skinMatrix * bindMatrix;
			vec4 skinnedNormal = skinMatrix * vec4( normal, 0.0 );
			vec3 objectNormal = skinnedNormal.xyz;
			vec3 transformedNormal = normalMatrix * objectNormal;
			vNormal = normalize( transformedNormal );
			vec4 skinVertex = bindMatrix * vec4( position, 1.0 );
			vec4 skinned = vec4( 0.0 );
			skinned += boneMatX * skinVertex * skinWeight.x;
			skinned += boneMatY * skinVertex * skinWeight.y;
			skinned += boneMatZ * skinVertex * skinWeight.z;
			skinned += boneMatW * skinVertex * skinWeight.w;
			skinned  = bindMatrixInverse * skinned;
			vec4 mvPosition = modelViewMatrix * skinned;
		#else
			vec4 mvPosition = modelViewMatrix * vec4( position, 1.0 );
		#endif
		
It works for Static meshes only.
In other words, for 
			vec4 mvPosition = modelViewMatrix * vec4( position, 1.0 );
only.

**Comment by mrdoob** [comment]
@tschw I think this is related to the glsl clean up you did. I don't think glsl should offer backwards compatibility (for performance reasons) but it would be nice to document this on the migrations page...
https://github.com/mrdoob/three.js/wiki/Migration

**Comment by tschw** [comment]
> @tschw I think this is related to the glsl clean up you did.

Possible, that is I don't categorically deny it.

That said, it looks more like a GLSL issue to me and I guess that the uniforms are missing for another reason: Our built-in shaders seem to use a vertex shader texture for most of the skinning data instead of uniforms.

**Comment by tschw** [comment]
Renderer code:

https://github.com/mrdoob/three.js/blob/master/src/renderers/WebGLRenderer.js#L1873-L1883

The `skeleton.useVertexTexture` path seems to be used by default. @RemusMar 's code OTOH seems to need the `else` path, so I guess `skeleton.useVertexTexture` must be set to `false` or the GLSL must be extended to handle both cases.

@mrdoob 
Do we have at least one example that enforces uniforms, so both variants are under test?

**Comment by mrdoob** [comment]
> @mrdoob 
> Do we have at least one example that enforces uniforms, so both variants are under test?

I don't think we do.

**Comment by RemusMar** [comment]
> The skeleton.useVertexTexture path seems to be used by default
	
		if ( skeleton ) {
			p_uniforms.setOptional( _gl, skeleton, 'boneMatrices' );
		}
		
didn't make any difference.
Please try to remember what did you "clean up" (and why).

**Comment by tschw** [comment]
@RemusMar 
Ha! I was just about to ask whether it helped.

Are you sure that `skeleton.boneMatrices` are actually being calculated when the rest of the code assumes it has to upload a texture? 

> Please try to remember what did you "clean up" (and why).

Please try to do exactly what I say, or at least try to reply to that part first, before making cynical comments ;).

**Comment by RemusMar** [comment]
> Please try to do exactly what I say, or at least try to reply to that part first, before making cynical comments ;).

Please read again (carefully) my first posted message.
You have VERY clean sources to fix what you did wrong.

**Comment by tschw** [comment]
@RemusMar 
Point is, your code is incomplete and not handling all cases that exist longer than `r75`. We just changed a default, which causes your incomplete code to break.

**Comment by RemusMar** [comment]
> Point is, your code is incomplete and not handling all cases that exist longer than r75.
> We just changed a default, which causes your incomplete code to break.

It was complete for over an year, but is not complete now ... ok, I can live with that.
Please tell me WHY did you change "a default"?
r76 does not provide fater rendering compared to r75.
On contrary !

**Comment by mrdoob** [comment]
Please, try to be less aggressive.

**Comment by RemusMar** [comment]
> Please, try to be less aggressive.

I'm here to help you cleaning the mess.

**Comment by tschw** [comment]
@RemusMar 
> Please tell me WHY did you change "a default"?

Please note that *I* did not change it (so I hope that is a plural "you"), but I'm kindly looking into your case :).
Although @mrdoob accused me of causing breakage, your issue is completely unrelated to my changes.  

> I'm here to help you cleaning the mess.

There are no reliable extension points for custom shaders (yet) and thus backward compatibility is not guaranteed. In fact, I'm one of the developers on this project who plan on changing this in the future, but we have to start where we're at, and so it may take a few more release cycles to actually get there.

Bottom line, please try a less negative tone. Nobody forces you to update. It's not that older versions of Three.js suddenly stop working once a new one is released.

**Comment by tschw** [comment]
Ok, my suspicion turns out is not the problem, you are passing an explicit `false` for `useVertexTexture`:

```javascript
lara = new THREE.SkinnedMesh( geometry, shadMat, false );
```

@RemusMar 
> You have VERY clean sources to fix what you did wrong.

Please include a non-minified build of Three.js that allows the use of a debugger.

**Comment by RemusMar** [comment]
> Please include a non-minified build of Three.js that allows the use of a debugger. 

http://necromanthus.com/Test/html5/Lara_shader_r76.html

**Comment by tschw** [comment]
@RemusMar 
Thanks for the quick response.

Renamed the vertex shader uniform `boneGlobalMatrices` to `boneMatrices` and runs.
Updated https://github.com/mrdoob/three.js/wiki/Migration .

**Comment by RemusMar** [comment]
> Renamed the vertex shader uniform boneGlobalMatrices to boneMatrices and runs. 

Indeed.
Now please tell me why did you rename it.

**Comment by mrdoob** [comment]
> Now please tell me why did you rename it.

Because life would be boring otherwise. Life won't be the same without @RemusMar being angry 😉

**Comment by mrdoob** [comment]
@tschw thanks for taking a look!

**Comment by RemusMar** [comment]
> Because life would be boring otherwise. Life won't be the same without @RemusMar being angry 

There are 3 possible explanations for breaking the backwards compatibility all the time:
1) poor programming
2) lack of vision
3) disrespect for customers and fans.

The choice is yours
cheers

**Comment by tschw** [comment]
@RemusMar wrote:
> Now please tell me why did you rename it.

It was a cleanup step toward the longer termed goal of having extension points to hook into the shading system instead of using fragile copypasta as we (extenders of the library) have to do now. IOW to get rid of the kind of stuff that causes the kind of trouble you are complaining about altogether.

Since the GLSL level is currently internal (and yes, that bugs me too) extenders have to adapt their code to internal changes. I do not think the change of an internal name should break any code, but there will be several more changes until we can make it happen.

@mrdoob wrote:
> Because life would be boring otherwise. Life won't be the same without @RemusMar being angry :wink:

I guess it's your way of saying: "Thanks for pulling up my code!" :)
